nuxeo-notification-stream-default/src/test/java/org/nuxeo/ecm/notification/TestDescendantChangesResolver.java

```diff
@@ -0,0 +1,173 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  Contributors:
+ *      Nuxeo
+ */
+
+package org.nuxeo.ecm.notification;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.failBecauseExceptionWasNotThrown;
+import static org.nuxeo.ecm.core.api.security.SecurityConstants.READ;
+import static org.nuxeo.ecm.notification.resolver.DescendantChangesResolver.FOLDERISH_ID_FIELD;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.IntStream;
+
+import javax.inject.Inject;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.api.PathRef;
+import org.nuxeo.ecm.core.api.security.ACE;
+import org.nuxeo.ecm.core.api.security.ACP;
+import org.nuxeo.ecm.notification.notifier.CounterNotifier;
+import org.nuxeo.ecm.platform.test.PlatformFeature;
+import org.nuxeo.ecm.platform.usermanager.UserManager;
+import org.nuxeo.runtime.stream.StreamHelper;
+import org.nuxeo.runtime.test.runner.Deploy;
+import org.nuxeo.runtime.test.runner.Features;
+import org.nuxeo.runtime.test.runner.FeaturesRunner;
+import org.nuxeo.runtime.test.runner.TransactionalFeature;
+
+@RunWith(FeaturesRunner.class)
+@Features({ NotificationFeature.class, PlatformFeature.class })
+@Deploy("org.nuxeo.ecm.platform.notification.stream.default")
+@Deploy("org.nuxeo.ecm.platform.collections.core")
+@Deploy("org.nuxeo.ecm.platform.notification.stream.default:OSGI-INF/default-contrib.xml")
+public class TestDescendantChangesResolver {
+
+    protected static final String USER_ID = "tmp-user";
+
+    @Inject
+    protected NotificationStreamCallback ncs;
+
+    @Inject
+    protected NotificationService ns;
+
+    @Inject
+    protected CoreSession session;
+
+    @Inject
+    protected UserManager userManager;
+
+    @Inject
+    protected TransactionalFeature txFeature;
+
+    @Before
+    public void before() {
+        DocumentModel user = userManager.getBareUserModel();
+        user.setPropertyValue(userManager.getUserIdField(), USER_ID);
+        userManager.createUser(user);
+
+        PathRef docRef = new PathRef("/");
+        ACP acp = session.getACP(docRef);
+        acp.getOrCreateACL().add(ACE.builder(USER_ID, READ).build());
+        session.setACP(docRef, acp, true);
+
+        waitAllAsync();
+    }
+
+    @After
+    public void after() {
+        waitAllAsync();
+    }
+
+    @Test(expected = NuxeoException.class)
```

Comment:
Kevin asked me once to check that the exception is properly thrown with a catch on the exact method rather than an expected exception globally to ensure the exception comes from the method you're testing and not from somewhere else.